security/pkg/stsservice/server/server.go

```diff
@@ -0,0 +1,233 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httputil"
+	"os"
+	"syscall"
+
+	"istio.io/istio/security/pkg/stsservice"
+	"istio.io/pkg/log"
+)
+
+const (
+	// tokenPath is url path for handling STS requests.
+	tokenPath = "/token"
+	// stsStatusPath is the path for dumping STS status.
+	stsStatusPath = "/stsStatus"
+	// urlEncodedForm is the encoding type specified in a STS request.
+	urlEncodedForm = "application/x-www-form-urlencoded"
+	// tokenExchangeGrantType is the required value for "grant_type" parameter in a STS request.
+	tokenExchangeGrantType = "urn:ietf:params:oauth:grant-type:token-exchange"
+	// subjectTokenType is the required token type in a STS request.
+	subjectTokenType = "urn:ietf:params:oauth:token-type:jwt"
+)
+
+var stsServerLog = log.RegisterScope("stsServerLog", "STS service debugging", 0)
+
+// error code sent in a STS error response. A full list of error code is
+// defined in https://tools.ietf.org/html/rfc6749#section-5.2.
+const (
+	// If the request itself is not valid or if either the "subject_token" or
+	// "actor_token" are invalid or unacceptable, the STS server must set
+	// error code to "invalid_request".
+	invalidRequest = "invalid_request"
+	// If the authorization server is unwilling or unable to issue a token, the
+	// STS server should set error code to "invalid_target".
+	invalidTarget = "invalid_target"
+)
+
+// Server watches HTTP requests for security token service (STS), and returns
+// token in response.
+type Server struct {
+	// tokenManager takes STS request parameters and generates tokens, and returns
+	// generated token to the STS server.
+	tokenManager stsservice.TokenManager
+	stsServer    *http.Server
+}
+
+// Config for the STS server.
+type Config struct {
+	LocalHostAddr string
+	LocalPort     int
+}
+
+// NewServer creates a new STS server.
+func NewServer(config Config, tokenManager stsservice.TokenManager) (*Server, error) {
+	s := &Server{
+		tokenManager: tokenManager,
+	}
+	mux := http.NewServeMux()
+	mux.HandleFunc(tokenPath, s.ServeStsRequests)
+	mux.HandleFunc(stsStatusPath, s.DumpStsStatus)
+	s.stsServer = &http.Server{
+		Addr:    fmt.Sprintf("%s:%d", config.LocalHostAddr, config.LocalPort),
+		Handler: mux,
+	}
+	go func() {
+		stsServerLog.Infof("Start listening on %s:%d", config.LocalHostAddr, config.LocalPort)
+		err := s.stsServer.ListenAndServe()
+		// ListenAndServe always returns a non-nil error.
+		stsServerLog.Errora(err)
+		notifyExit()
```

Comment:
why not just return?

Reply:
the only reason pilot agent has this is because it has a /quitquitquit endpoint and even that was probably a bad implementation

Reply:
Good advice, notifyExit() is removed. Thanks.